README.md
# README

The **Ckmeans.1d.dp** package provides fast, optimal, and reproducible univariate clustering by dynamic programming. It is practical to cluster millions of sample points with optional weights in seconds on a typical desktop computer using a single processor core.

Four types of problem including univariate $k$-means, $k$-median, $k$-segments, and multi-channel weighted $k$-means are solved with guaranteed optimality and reproducibility. The core algorithm minimizes the (weighted) sum of within-cluster distances using respective metrics. Its advantage over heuristic clustering in efficiency and accuracy is increasingly pronounced as the number of clusters $k$ increases. Weighted $k$-means can also optimally segment time series to perform peak calling. An auxiliary function generates histograms that are adaptive to patterns in data. This package provides a powerful set of tools for univariate data analysis with guaranteed optimality, efficiency, and reproducibility.

The Ckmeans.1d.dp algorithm clusters (weighted) univariate data given by a numeric vector $x$ into $k$ groups by dynamic programming (Wang and Song, 2011). It guarantees the optimality of clustering---the total of within-cluster sums of squares is always the minimum given the number of clusters $k$. In contrast, heuristic univariate clustering algorithms may be non-optimal or inconsistent from run to run. As unequal non-negative weights are supported for each point, the algorithm can also segment a time course using the time points as input and the values at each time point as weight. Utilizing the optimal clusters, a function can generate histograms adaptive to patterns in data.

Apart from the time for sorting $x$, the default weighted univariate clustering algorithm takes a runtime of $O(kn)$, linear in both sample size $n$ and the number of clusters $k$, using a new divide-and-conquer strategy integrating a previous theoretical result on matrix search (Aggarwal et al., 1987) and a novel in-place search space reduction method. The space complexity is $O(kn)$. This method is numerically stable.

This package provides a powerful alternative to heuristic clustering and also new functionality for weighted clustering, segmentation, and peak calling with guaranteed optimality.


NEWS
NEWS

Version 4.2.2
  Changes
  2018-09-21
  1. Modified the package to use Rcpp interface instead of the
     old-style C interface.

Version 4.2.1
  Changes
  2017-06-10
  1. Added a new vignette "Tutorial: Linear weight scaling in cluster analysis".
  2. Re-organized manuals and updated documentation.

Version 4.2.0
  Changes
  2017-05-29
  1. Increased log likelihood calculation to long double
     precision in C++ function weighted_select_levels.cpp.
  2. Replaced std::accumulate() function by for-loop addition in
     C++ function weighted_select_levels.cpp. This resolved a
     numerical overflow issue when the weight values are large.
  3. Now R function plotBIC() automatically adjusts the "k*="
     text position, so that the text label is placed entirely
     within the BIC curve area and will not extend into the
     figure margin.
  4. The cluster size has been changed from integer to double to
     accomodate weighted cluster size in both R and C++ code.
  5. Force any weight vector to be equal weight in new R
     function Ckmedian.1d.dp.
  6. Introduced S3 methods print and plot for Ckmedian.1d.dp
     and Cksegs.1d.dp objects.

  2017-03-18
  1. Introduced Cksegs.1d.dp() function for k-segments clustering
     of y with or without x. Only method="quadratic" guarantees
     optimality.

  2. Expanded k-median clustering to work with all possible
     methods. Only unweighted solution guarantees optimality.

Version 4.1.0
  Changes
  2017-03-02
  1. Introduced function Ckmedian.1d.dp() for k-median unweighted
     clustering.

Version 4.0.2
  Changes
  2017-02-16
  1. Fixed symbol encoding used in NEWS.
  2. Updated documentation.

Version 4.0.1
  Changes
  2017-02-16
  1. Fixed a warning message in the use of 'R_registerRoutines' and
     'R_useDynamicSymbols'.
  2. Fixed a memory leak issue: invalid read of size 8.

Version 4.0.0
  Changes
  2017-02-11
  1. Removed some examples for future use.

Version 3.4.15
  1. Minor changes.

Version 3.4.14
  Changes
  2017-01-03
  1. Minor changes in documentation files.

  2017-01-02
  1. Changed package title to "Optimal and Fast Univariate Clustering".

  2016-12-27
  1. When the input vector x is empty, function Ckmeans.1d.dp now generates a
     warning message instead of stops on error. Ckmeans.1d.dp.R is modified.
  2. Print out appropriate warning messages when input x does not have an
     appropriate type. ahist.R is modified.

Version 3.4.13
  Changes
  2016-12-19
  1. Revised the comparison function in sorting so that the code can be
     compiled by C++98, as requested by a user.

  2016-12-06
  1. Expanded the ahist() function to support weighted adaptive histogram

  2016-12-04
  1. Expanded the vignette of adaptive histograms to a tutorial.
  2. Expanded the vignette of optimal univariate k-means clustering to a tutorial.
  3. Update the time course example in Ckmeans.1d.dp function

  2016-10-22
  1. Added a vignette to visualize examples of adaptive histograms.
  2. Added a vignette to visualize examples of optimal univariate k-means
  clustering.

  2016-10-21
  1. Added an equal bin width histogram example to contrast with the adaptive
  histogram.

  2016-10-16
  1. Moved ahist() function from visualize.R to a new R file ahist.R.

  2016-10-15
  1. Added a breaks argument to ahist() so as use default graphics::hist() but
     with the capacity to add sticks to the histogram generated.
  2. Added a skip.empty.bin.color argument to ahist() to gain more control over
     colors of the histogram bars.

  2016-10-12
  1. Added a data argument to ahist() to provide raw data for visualization.
  2. Allow x to ahist() to be an object of the class "Ckmeans.1d.dp" to avoid
     recomputing the clustering if it has already been done. This requires the
     data for clustering to be provided via the data argument.

  2016-10-11
  1. Added an argument add.sticks=TRUE to ahist() to turn on or off the sticks
  just above the horizontal axis.
  2. Added an argument style to ahist() for different styles of adaptive
  histogram.

  2016-10-01
  1. Added a new function plot() to visualize the clusters.
  2. Added a new function plotBIC() to show the Bayesian information criterion
  as a function of number of clusters.

  2016-09-30
  1. Updated examples of ahist().
  2. Added sticks to ahist() to show the original input data.

  2016-09-27
  1. Fixed ahist() when there is only a single bin detected.
  2. Made ahist() run much faster than the previous version.
  3. Updated previous examples and added more examples to illustrate
     the use of ahist() better.

  2016-09-25
  1. Introduced a new function ahist() to generate adaptive histograms
     corresponding to the optimal univariate k-means clustering.

  2016-09-24
  1. Known issue: loglinear option may generate optimal clustering different
     from linear and quadratic.
  2. The default k estimation method is updated. Updated number of cluster k
     estimation. The main difference is when there are duplicates in the data.
     Otherwise, the estimated k would be the same with previous versions. Added
     an argument estimate.k in fiction Ckmeans.1d.dp() to use the BIC method in
     version 3.4.12 or earlier to estimated k for compatibility.

Version 3.4.12
  Changes
  2016-08-20
  1. The weighted univariate k-means now runs in O(kn), down from O(kn^2).
     This is a result of integrating weighted and unweighted k-means
     clustering into a unified dynamic programming function without sacrificing
     performance. This also fixed a bug in the previous loglinear-time weighted
     k-means implementation.

Version 3.4.9
  Changes
  2016-07-19
  1. If the input array is already sorted, sorting is not performed again.
  2. Added an option method to select either the linear or loglinear algorithm.

  2016-07-16
  1. Implemented linear recursive algorithm based on the method described in
  (Aggarwal et al., 1987)

Version 3.4.8
  Changes
  2016-06-01

R/Ckmeans.1d.dp.R
## Ckmeans.1d.dp.R
##
##   Ckmeans.1d.dp()-----Interface funcition to call C++ version kmeans.1d.dp()
##
##   Haizhou Wang
##   Computer Science Department
##   New Mexico State University
##   hwang@cs.nmsu.edu
##
##   Created: Oct 20, 2009
##
## Modified:
#    May 17, 2016. MS
#    September 25, 2016. MS. Introduced function ahist()
#    September 21, 2018. Joe Song. Changed from C to Rcpp interface


##Ckmeans.1d.dp : function which implement optimal one-dimensional clustering
## x is one-dimensional input vector
## k indicates cluster level
cluster.1d.dp <- function( x, k, y, method, estimate.k, criterion, xname, yname )
{
  if(is.null(k)) {
    k <- 1: min( 9, length(x) )
  } else {
    k <- as.integer(ceiling(k))
  }

  if(length(k) > 1) {
    k.min <- min(k)
    k.max <- max(k)
  } else {
    k.min <- k
    k.max <- k
  }

  ##Check to see if k is less than 0.
  ##If k is less than 0, stop and exit.
  if( k.max <= 0 ) {
    stop ("Can NOT classify vector into 0 or less cluster\n")
  }

  n.unique <- length(unique(x))
  ##Check to see if cluster level bigger than the unique number of the input vector
  ##If k is bigger than the unique number of the input vector,
  ##force k set to the number of unique number of input.

  if(n.unique == 0) {
    warning(paste("Input vector", deparse(substitute(x)), "is empty!\n"))
  } else {
    if(n.unique < k.min) {
      warning("Min number of clusters is greater than the unique number of elements in\n",
              "the input vector, both k.min and k.max are set to the number of\n",
              "unique number of input values.\n")
      k.min <- n.unique
      k.max <- n.unique
    } else if (n.unique >= k.min && n.unique < k.max) {
      warning("Max number of clusters is greater than the unique number of\n",
              "elements in the input vector, and k.max is set to the number of\n",
              "unique number of input values.\n")
      k.max <- n.unique
    }
  }

  ##Form data which will be passed to external C++ function.
  clusters <- vector("integer", length(x))
  center <- vector("double", k.max)
  withinss <- vector("double", k.max)
  size <- vector("double", k.max) # vector("integer", k.max)
  BIC <- vector("double", k.max-k.min+1)

  #Call external C++ function
  if(criterion == "L2") {

    # result <- .C("Ckmeans_1d_dp", PACKAGE="Ckmeans.1d.dp",
    #              data=as.double(x), length=as.integer(length(x)),
    #              weight=as.double(y), weight_length=as.integer(length(y)),
    #              Kmin=as.integer(k.min), Kmax=as.integer(k.max),
    #              cluster=as.integer(clusters), centers=as.double(center),
    #              withinss=as.double(withinss), size=as.double(size), # size=as.integer(size),
    #              BIC=as.double(BIC),
    #              estimate.k=as.character(estimate.k),
    #              method=as.character(method))

    result <- Ckmeans_1d_dp(
                 as.double(x), as.integer(length(x)),
                 as.double(y), as.integer(length(y)),
                 as.integer(k.min), as.integer(k.max),
                 as.integer(clusters), as.double(center),
                 as.double(withinss), as.double(size), # size=as.integer(size),
                 as.double(BIC),
                 estimate.k,
                 method)

    class <- "Ckmeans.1d.dp"
  } else if (criterion == "L1") {

    # result <- .C("Ckmedian_1d_dp", PACKAGE="Ckmeans.1d.dp",
    #              data=as.double(x), length=as.integer(length(x)),
    #              weight=as.double(y), weight_length=as.integer(length(y)),
    #              Kmin=as.integer(k.min), Kmax=as.integer(k.max),
    #              cluster=as.integer(clusters), centers=as.double(center),
    #              withinss=as.double(withinss), size=as.double(size), # size=as.integer(size),
    #              BIC=as.double(BIC),
    #              estimate.k=as.character(estimate.k),
    #              method=as.character(method))

    result <- Ckmedian_1d_dp(
      as.double(x), as.integer(length(x)),
      as.double(y), as.integer(length(y)),
      as.integer(k.min), as.integer(k.max),
      as.integer(clusters), as.double(center),
      as.double(withinss), as.double(size), # size=as.integer(size),
      as.double(BIC),
      estimate.k,
      method)

    class <- "Ckmedian.1d.dp"
  } else if(criterion == "L2Y") {

    # result <- .C("Cksegs_1d_dp", PACKAGE="Ckmeans.1d.dp",
    #              data=as.double(x), length=as.integer(length(x)),
    #              weight=as.double(y), weight_length=as.integer(length(y)),
    #              Kmin=as.integer(k.min), Kmax=as.integer(k.max),
    #              cluster=as.integer(clusters), centers=as.double(center),
    #              withinss=as.double(withinss), size=as.double(size), # size=as.integer(size),
    #              BIC=as.double(BIC),
    #              estimate.k=as.character(estimate.k),
    #              method=as.character(method))

    result <- Cksegs_1d_dp(
      as.double(x), as.integer(length(x)),
      as.double(y), as.integer(length(y)),
      as.integer(k.min), as.integer(k.max),
      as.integer(clusters), as.double(center),
      as.double(withinss), as.double(size), # size=as.integer(size),
      as.double(BIC),
      estimate.k,
      method)

    class <- "Cksegs.1d.dp"
  }

  if(length(result$cluster) > 0) {
    k.opt <- max(result$cluster) # length(unique(result$cluster))
  } else {
    k.opt <- 0
  }

  if(n.unique > 0) {
    if(k.min < k.max) {
      if (k.opt == k.min && k.min != 1) {
        warning("Min number of clusters used. Consider decreasing k!\n")
      } else if(k.opt == k.max && k.max != length(x)) {
        warning("Max number of clusters used. Consider increasing k!\n")
      }
    }
  }

  if(length(y) == length(x) && sum(y) != 0 && criterion != "L2Y") {
    totss <- sum(y * (x - sum(as.numeric(x * y)) / sum(y))^2)
  } else if(criterion == "L2Y") {
    totss <- sum((y - sum(as.numeric(y)) / length(y))^2)
  } else {
    # totss <- sum(scale(x, scale=FALSE)^2) scale function is VERY SLOW!
    totss <- sum((x - sum(as.numeric(x)) / length(x))^2)
  }

  tot.withinss <- sum(result$withinss[1:k.opt])
  betweenss <- totss - tot.withinss
  BIC <- result$BIC
  names(BIC) <- paste0("k=", k.min : k.max)

  r <- structure(
    list(
      cluster = result$cluster, centers = result$centers[1:k.opt],
      withinss = result$withinss[1:k.opt], size = result$size[1:k.opt],
      totss = totss, tot.withinss = tot.withinss, betweenss = betweenss,
      BIC = BIC, xname=xname, yname=yname
    ),
    class = class)

  return( r )
} ##end of cluster.1d.dp()


Ckmeans.1d.dp <- function( x, k=c(1,9), y=1,
                           method=c("linear", "loglinear", "quadratic"),
                           estimate.k=c("BIC", "BIC 3.4.12") )
{
  method <- match.arg(method)
  estimate.k <- match.arg(estimate.k)
  cluster.1d.dp(x, k, y, method, estimate.k, "L2",
                deparse(substitute(x)), deparse(substitute(y)))
}

Ckmedian.1d.dp <- function( x, k=c(1,9), y=1,
                            method=c("linear", "loglinear", "quadratic"),
                            estimate.k=c("BIC", "BIC 3.4.12") )
{
  method <- match.arg(method)
  estimate.k <- match.arg(estimate.k)

  if(length(y) > 1) {
    warning("Ckmedian.1d.dp() ignores weight vector y and uses equal weight instead!")
  }

  cluster.1d.dp(x, k, y=1, method, estimate.k, "L1",
                deparse(substitute(x)), deparse(substitute(y)))
}

Cksegs.1d.dp <- function(y, k=c(1,9), x=seq_along(y),
                         method=c("quadratic", "linear", "loglinear"),
                         estimate.k=c("BIC", "BIC 3.4.12") )
{
  method <- match.arg(method)
  estimate.k <- match.arg(estimate.k)
  cluster.1d.dp(x, k, y, method, estimate.k, "L2Y",
                deparse(substitute(x)), deparse(substitute(y)))
}
